src/integer_value.cpp

```diff
@@ -116,27 +117,26 @@ ValuePtr IntegerValue::mul(Env *env, ValuePtr arg) {
     }
     arg.assert_type(env, Value::Type::Integer, "Integer");
 
+    if (is_zero() || arg->as_integer()->is_zero())
+        return new IntegerValue { 0 };
+
     if (arg.is_bignum()) {
         auto other = arg->as_integer();
         auto result = to_bignum() * other->to_bignum();
         return new BignumValue { result };
     }
 
-    nat_int_t result = to_nat_int_t() * arg.to_nat_int_t();
-    bool overflowed = false;
-    bool same_sign = (to_nat_int_t() ^ arg.to_nat_int_t()) >= 0;
-    if (!same_sign && result > 0)
-        overflowed = true;
-    if (same_sign && result < 0)
-        overflowed = true;
+    auto ll_this = to_nat_int_t();
+    auto ll_arg = arg.to_nat_int_t();
 
-    if (overflowed) {
-        auto other = arg->as_integer();
-        auto result = to_bignum() * other->to_bignum();
-        return new BignumValue { result };
+    auto min_fraction = (NAT_MIN_FIXNUM - (NAT_MIN_FIXNUM % ll_arg)) / ll_arg;
+    auto max_fraction = (NAT_MAX_FIXNUM - (NAT_MAX_FIXNUM % ll_arg)) / ll_arg;
+    if (
+        (ll_this > 0 && ll_arg > 0 && max_fraction <= ll_this) || (ll_this > 0 && ll_arg < 0 && min_fraction <= ll_this) || (ll_this < 0 && ll_arg > 0 && min_fraction >= ll_this) || (ll_this < 0 && ll_arg < 0 && max_fraction >= ll_this)) {
```

Comment:
This looks a bit different to what [ruby ](https://github.com/ruby/ruby/blob/cb4e2cb55a59833fc4f1f6db2f3082d1ffcafc80/internal/bits.h#L102)does. Basically they do not need this modulo operation. Can you explain to me why it's there?

Reply:
yeah, I think I just dumbly added that thinking you could have an issue when arg = NAT_MAX_FIXNUM, but it doesn't seem to be the case and really doesn't make much sense :) Thanks!